### PR TITLE
feat(grainlsp): Implement inlay hints for module includes

### PR DESCRIPTION
**Pull request by marcusroberts**
Establish framework for adding inlay hints with the LSP.

Show hint for name of module included (useful for where the module name differs from what you might guess from the included file name)

### CODE REVIEW

**Comment by marcusroberts** [comment]
This is a minimal implementation of inlay hints.   We can easily add any we want when iterating through the typed tree.
For better performance this should be scoped to the range specified by the editor in the request.

File: compiler/src/language_server/inlayhint.re [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          let stmt_loc = stmt.ttop_loc;
+          let stmt_end = stmt_loc.loc_end;
+
+          let p: Protocol.position = {
+            line: stmt_end.pos_lnum - 1,
+            character: stmt_end.pos_cnum - stmt_end.pos_bol + 1 + 1,
+          };
+
+          let r: ResponseResult.inlay_hint = {
+            label: " : " ++ name,
```
Comment (spotandjake) [inline compiler/src/language_server/inlayhint.re:63]: Question on if the leading space should be removed making it `": " ++ name`, when you showed the picture in discord it looked a little off to me having the leading space.

**Review by spotandjake — COMMENTED** [review]
Looks great, I had one comment but its really a style thing and either way I think this looks good.

File: compiler/src/language_server/protocol.rei [hunk 2]
```diff
@@ -1,5 +1,12 @@
 open Grain_utils;
 
+//https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintOptions[@deriving yojson]
```
Comment (phated) [inline compiler/src/language_server/protocol.rei:3]:
```suggestion
//https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintOptions
```

File: compiler/src/language_server/inlayhint.re [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+open Compile;
+open Grain_parsing;
+open Grain_utils;
+open Grain_typed;
+open Grain_diagnostics;
+open Sourcetree;
+open Lsp_types;
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintParams
+
```
Comment (phated) [inline compiler/src/language_server/inlayhint.re:11]:
```suggestion
// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintParams
```

**Review by phated — CHANGES_REQUESTED** [review]
This looks great! 🎉 I hope the patterns were easy to follow. I had a few nit suggestions, but otherwise looks good.

File: compiler/src/language_server/inlayhint.re (hunk 1 above)
Comment (marcusroberts) [reply compiler/src/language_server/inlayhint.re:63]: I think you are right for when it's in a hint.

File: compiler/src/language_server/inlayhint.re (hunk 3 above)
Comment (marcusroberts) [reply compiler/src/language_server/inlayhint.re:11]: I've stared at this for ages, what's the code change?

File: compiler/src/language_server/inlayhint.re [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      include TypedtreeIter.DefaultIteratorArgument;
+      let enter_toplevel_stmt = (stmt: toplevel_stmt) => {
+        switch (stmt.ttop_desc) {
+        | TTopInclude(inc) =>
+          let path = inc.tinc_path;
+          let name =
+            switch (path) {
+            | PIdent(i) => i.name
+            | _ => ""
+            };
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/language_server/inlayhint.re]: It does indeed!

File: compiler/src/language_server/protocol.rei (hunk 2 above)
Comment (marcusroberts) [reply compiler/src/language_server/protocol.rei]: Fixed.

File: compiler/src/language_server/inlayhint.re (hunk 3 above)
Comment (phated) [reply compiler/src/language_server/inlayhint.re:11]: It's just removing the space so the comment is attached to a node.

File: compiler/src/language_server/inlayhint.re [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+open Compile;
+open Grain_parsing;
+open Grain_utils;
+open Grain_typed;
+open Grain_diagnostics;
+open Sourcetree;
+open Lsp_types;
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintParams
+
```
Comment (phated) [inline compiler/src/language_server/inlayhint.re:11]:
```suggestion
// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintParams
```

File: compiler/src/language_server/inlayhint.re (hunk 3 above)
Comment (marcusroberts) [reply compiler/src/language_server/inlayhint.re]: Ah, done.

**Review by phated — APPROVED** [review]

**Pull request merged**